torch/autograd/gradcheck.py

```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
-        if is_tensor_like(inp) and inp.requires_grad:
-            if not (inp.dtype == torch.float64 or inp.dtype == torch.complex128):
-                warnings.warn(
-                    f'Input #{idx} requires gradient and '
-                    'is not a double precision floating point or complex. '
-                    'This check will likely fail if all the inputs are '
-                    'not of double precision floating point or complex. ')
-            content = inp._values() if inp.is_sparse else inp
-            # TODO: To cover more problematic cases, replace stride = 0 check with
-            # "any overlap in memory" once we have a proper function to check it.
-            if content.layout is not torch._mkldnn:  # type: ignore
-                if not all(st > 0 or sz <= 1 for st, sz in zip(content.stride(), content.size())):
-                    raise RuntimeError(
-                        'The {}th input has a dimension with stride 0. gradcheck only '
-                        'supports inputs that are non-overlapping to be able to '
-                        'compute the numerical gradients correctly. You should call '
-                        '.contiguous on the input before passing it to gradcheck.')
-            any_input_requiring_grad = True
-            inp.retain_grad()
-    if not any_input_requiring_grad:
-        raise ValueError(
-            'gradcheck expects at least one input tensor to require gradient, '
-            'but none of the them have requires_grad=True.')
+    if not check_inputs(fail_test, tupled_inputs, check_sparse_nnz):
+        return False
 
     func_out = func(*tupled_inputs)
-    output = _differentiable_outputs(func_out)
-
-    if not output:
-        for i, o in enumerate(func_out):
-            def fn(input):
-                return _as_tuple(func(*input))[i]
-            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
-            for n in numerical:
-                if torch.ne(n, 0).sum() > 0:
-                    return fail_test('Numerical gradient for function expected to be zero')
-        return True
+    outputs = _differentiable_outputs(func_out)
 
-    for i, o in enumerate(output):
-        if not o.requires_grad:
-            continue
+    check_outputs(outputs)
+
+    if not outputs:
+        return check_no_differentiable_outputs(fail_test, func, tupled_inputs, func_out, eps)
 
+    for i, o in enumerate(outputs):
         def fn(input):
             return _as_tuple(func(*input))[i]
 
-        analytical, reentrant, correct_grad_sizes, correct_grad_types = get_analytical_jacobian(tupled_inputs,
-                                                                                                o,
-                                                                                                nondet_tol=nondet_tol)
+        analytical, failed = check_analytical_jacobian(tupled_inputs, o, nondet_tol, 1.0, check_batched_grad,
```

Comment:
The name here is a bit ambiguous. What is this actually checking? Maybe "check_analytical_jacobian_attributes" makes more sense (because the check runs tests for dtypes and sizes among other things)